File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == nullptr) {
            return head;
        }
        Node* copy = head;
        unordered_map<Node*, Node*> mp;

        while (copy) {
            mp[copy] = new Node(copy->val);
            copy = copy->next;
        }
        copy = head;
        while(copy) {
            Node* copyNext = copy->next;
            Node* copyRandom = copy->random;
            mp[copy]->next = mp[copyNext]; 
            mp[copy]->random = mp[copyRandom];
            copy = copy->next;
        }
        return mp[head];
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) {
            return head;
        }
        Node copy = head;
        Map<Node, Node> map = new HashMap<>();
        while (copy != null) {
            map.put(copy, new Node(copy.val));
            copy = copy.next;
        }
        copy = head;
        while (copy != null) {
            Node copyNext = copy.next;
            Node copyRandom = copy.random;
            map.get(copy).next = map.get(copyNext); 
            map.get(copy).random = map.get(copyRandom);
            copy = copy.next;
        }
        return map.get(head);
    }
}